### ISSUE TITLE
Remove `BaseError` implementation hack that depends on bad closure-compiler behavior.

### ISSUE THREAD

**Issue by brad4d** [angular/angular#12575]
**I'm submitting a ...**  (check one with "x")

```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
https://github.com/angular/angular/blob/745e10e6d2ea9097b7ec650ae54cea91e3d193f2/modules/%40angular/facade/src/errors.ts#L23

Depends on incorrect current behavior of closure-compiler which causes `super()` to evaluate to
a newly created `Error` object when extending `Error`.

**Expected behavior**
I'll be fixing the compiler so inheritance from `Error` works as expected and there's no need for the
hack being done here.
google/closure-compiler#2102

Once that's done, I believe the definition of `BaseError` can be changed to just this.

``` javascript
class BaseError extends Error {}
```

**Minimal reproduction of the problem with instructions**
n/a

**What is the motivation / use case for changing the behavior?**
I want to change the closure-compiler behavior to match the spec.
google/closure-compiler#2109

**Please tell us about your environment:**
n/a
- **Angular version:** 2.0.X
- **Browser:** [all ]
- **Language:** [TypeScript X.X ]
- **Node (for AoT issues):** `node --version` = n/a

**Comment by brad4d** [comment]
FYI @mprobst

**Comment by mprobst** [comment]
`class BaseError extends Error {}` looks a lot nicer than what we have right now. Will it work when executing in the browser directly though? How would we write a `constructor()` in that code that correctly calls `super`?

**Comment by mprobst** [comment]
/CC @mhevery who understands this ES6 corner case better than I do.

**Comment by brad4d** [comment]
@mprobst inheriting from Error works correctly with true ES6 classes, so if the code can run in the browser without transpilation, then inheritance from Error will work exactly as it should.

If you wanted to extend Error with a class that provided a constructor, here's how that would look.
I've also thrown in setting the `name` field on the prototype.

```javascript
class WidgetError extends Error {
  /**
   * @param {string} message
   * @param {Widget=} widget involved in the error
   */
  constructor(message, widget) {
    if (widget) {
      message += `: ${widget.toString()}`;
    }
    super(message);
  }
}
WidgetError.prototype.name = 'WidgetError';
```

**Comment by mhevery** [comment]
@brad4d Here is the issue: https://plnkr.co/edit/vHIeNM89RFbORk96GhB0?p=preview

As you can see the `new WidgetError` will not have a `stack` property and so will fail to be useful as an Error.

**Comment by tbosch** [comment]
@mhevery can this be closed then?

**Comment by brad4d** [comment]
@mhevery, please help me understand your example.

I see from the output stack that some form of compilation seems to have been done on the source code.
Is closure-compiler involved here? If so, which version?

Here's what it looks like to me.
1.  You're directly extending Error.
2.  Closure-compiler is being used, and it's an older version than my recent change to fix the handling of extending Error.
3.  The output code predictably doesn't work correctly, because my fix isn't in place yet.

Here's the github commit that fixes closure-compiler to make extending Error work.
https://github.com/google/closure-compiler/commit/abe8cddccaa548c8e9e46dd6406dd455cff9d978

I would expect it to be in the official external release that comes in November.

Please note also this line in that change that points out the problem being tracked in this bug.
https://github.com/google/closure-compiler/commit/abe8cddccaa548c8e9e46dd6406dd455cff9d978#diff-b9c32e889ab2b4502e1d1862f2df5a69R143

**Comment by brad4d** [comment]
Ah, I see my suggested temporary fix for this was merged in.

https://github.com/brad4d/angular/blob/0f5f15dc1651c2d1d3cae30c1ab9ea6f63c7ac90/modules/%40angular/facade/src/errors.ts#L26

Great.

I think this issue should stay open until

- closure-compiler release includes correct handling extending Error (ETA November release)
- That version of closure-compiler or later is used by angular.
- The whole hack of delegating to an internal Error object is removed.

**Comment by brad4d** [comment]
@mprobst & @mhevery 
Is angular always used with closure-compiler?
Behavior of extending Error is not correctly handled by Babel or Traceur.

**Comment by mprobst** [comment]
@brad4d I think most people use Angular 2 with the TypeScript compiler transpiling to ES5, and then just using webpack + uglify for bundling and (some) minification.

**Comment by brad4d** [comment]
So, I guess the question is, how does TypeScript manage the transpilation
when extending Error classes?
I wonder why I even tripped over this, if the code is expected to be
transpiled to ES5 by TypeScript before closure-compiler ever sees it.

**Comment by mprobst** [comment]
Users within Google all (I hope) use Closure Compiler, and pass ES6 to Closure Compiler. Users outside of Google mostly don't use CC, and use TS to down-transpile to ES5. TS emits the literal JS code you're writing into the constructor, see [here](http://www.typescriptlang.org/play/#src=class%20MyError%20extends%20Error%20%7B%0D%0A%20%20%20%20constructor()%20%7B%0D%0A%20%20%20%20%20%20%20%20let%20ex%20%3D%20super()%20as%20any%3B%0D%0A%20%20%20%20%20%20%20%20ex.stack%3B%0D%0A%20%20%20%20%7D%0D%0A%7D).

**Comment by brad4d** [comment]
In that case the hack will have to stay in place until TypeScript can correctly handle extending Error when transpiling to ES5.

**Comment by alexeagle** [comment]
Update: TS 2.1 has changed the behavior here, see
https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work

**Comment by mhevery** [comment]
closed by 0f5f15dc1651c2d1d3cae30c1ab9ea6f63c7ac90
